---
title: "3D pipeline"
author: "Wenbin Guo"
date: "2 November 2018"
output:
  html_document:
    code_folding: show
    fig_caption: yes
    highlight: textmate
    theme: default  
    toc: yes
    toc_depth: 4
    toc_float:
      collapsed: no
      smooth_scroll: yes
  pdf_document:
    toc: yes
    toc_depth: '4'
  word_document:
    toc: no
    toc_depth: '4'
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = T,eval = F,message = T, warning = F,
                      fig.width=12,fig.height=9,fig.align = "center")
```

## Introduction
Instead of mouse click analysis in the <a href="xxx" target="_blank">3D RNA-seq App</a>, advanced R users also can use command line to do the same analysis. This file includes step-by-step codings, which correspond to the steps in the 3D RNA-seq App. Users can refer to the details in the manual:xxx.

## Load R package
Assume all the required R packages are installed.
```{r,message = F}
.libPaths('D:/R/win_library')
# library(ThreeDRNAseq)
## app.R ##
library(shiny)
library(shinydashboard)
library(rhandsontable)
library(shinyFiles)
library(tximport)
library(edgeR)
library(limma)
library(plotly)
library(RUVSeq)
library(eulerr)
library(gridExtra)
library(grid)
library(ComplexHeatmap)
library(Gmisc)
options(stringsAsFactors=F)

sourceDir <- function(path, trace = TRUE, ...) {
  for (nm in list.files(path, pattern = "[.][RrSsQq]$")) {
    #if(trace) cat(nm,":")
    source(file.path(path, nm), ...)
    #if(trace) cat("/n")
  }
}
sourceDir('R')

################################################################################
### Set folder names to save results
##----->> save results to local folders
data.folder <- 'data' # for .RData format results
result.folder <- 'result' # for 3D analysis results in csv files
figure.folder <- 'figure' # for figures

### Set the input data folder
##----->> input data folder of samples.csv, mapping.csv, contrast.csv, etc.
input.folder <- 'examples/Arabidopsis_cold'
```

## Data generateion
Use the <a href="http://bioconductor.org/packages/release/bioc/html/tximport.html" target="_blank">tximport</a> R package to generate the transcript and gene level read counts and TPMs (transcript per million reads).

```{r}
################################################################################
##----->> Sample information, e.g. conditions, bio-reps, seq-reps, abundance paths, etc.
samples <- read.csv(paste0(input.folder,'/samples.csv'))
samples$condition <- gsub(pattern = '_','.',samples$condition)
samples$sample.name <- paste0(samples$condition,'_',samples$brep,'_',samples$srep)
save(samples,file=paste0(data.folder,'/samples.RData'))
data.files <-  as.character(samples$path) #Complete path of abundance, e.g. quant.sf files of salmon.

##----->> Transcript-gene mapping
mapping <- read.csv(paste0(input.folder,'/mapping.csv'))
rownames(mapping) <- mapping$TXNAME
save(mapping,file=paste0(data.folder,'/mapping.RData'))

################################################################################
##----->> Generate gene expression
abundance.from <- 'salmon' # abundance generator
countsFromAbundance <- 'lengthScaledTPM' # method to generate expression in tximport
##
txi_genes <- tximport(data.files,dropInfReps = T,
                      type = abundance.from, tx2gene = mapping,
                      countsFromAbundance = countsFromAbundance)

## give colunames to the datasets
colnames(txi_genes$counts)<- samples$sample.name
colnames(txi_genes$abundance)<-samples$sample.name
colnames(txi_genes$length)<-samples$sample.name

## save the data
write.csv(txi_genes$counts,file=paste0(result.folder,'/counts_genes.csv'))
write.csv(txi_genes$abundance,file=paste0(result.folder,'/TPM_genes.csv'))
save(txi_genes,file=paste0(data.folder,'/txi_genes.RData'))

################################################################################
##----->> Generate transcripts expression
txi_trans<- tximport(data.files, type = abundance.from, tx2gene = NULL,
                     countsFromAbundance = countsFromAbundance,
                     txOut = T,dropInfReps = T)

## give colunames to the datasets
colnames(txi_trans$counts)<- samples$sample.name
colnames(txi_trans$abundance)<-samples$sample.name
colnames(txi_trans$length)<-samples$sample.name

## save the data
write.csv(txi_trans$counts,file=paste0(result.folder,'/counts_trans.csv'))
write.csv(txi_trans$abundance,file=paste0(result.folder,'/TPM_trans.csv'))
save(txi_trans,file=paste0(data.folder,'/txi_trans.RData'))

################################################################################
genes_counts=txi_genes$counts
trans_counts=txi_trans$counts

```

## Data pre-processing

### Step 1: Merge sequencing replicates
If the RNA-seq data includes sequencing replicates (i.e. the same biological replicate/library is sequenced multiple times), the sequencing replicates are merged to increase the library depths, since they do not provided useful information for biological/technical variability.

```{r}
##If no sequencing replicates, genes_counts and trans_counts remain the same by 
##runing the command
idx <- paste0(samples$condition,'_',samples$brep)
genes_counts <- sumarrays(genes_counts,group = idx)
trans_counts <- sumarrays(trans_counts,group = idx)

## update the samples
samples_new <- samples[samples$srep==samples$srep[1],]
save(samples_new,file=paste0(data.folder,'/samples_new.RData'))
```

### Step 2: Filter low expression transcripts/genes
To filter the low expressed transcripts across the samples, the expression mean-variance trend is used to determine the CPM (count per million reads) cut-off for low expression and the number of samples to cut. A gene is expressed if any of its transcript is expressed. Please see the details in the 3D RNA-seq App user manual. 

```{r}
################################################################################
##----->> Do the filters
cpm.cut <- 1
sample.n.cut <- 1
target_high <- counts.filter(counts = trans_counts,
                             mapping = mapping,
                             cpm.cut = cpm.cut,
                             sample.n = cpm.cut,
                             Log=F)
save(target_high,file=paste0(data.folder,'/target_high.RData'))

################################################################################
##----->> Mean-variance plot

## transcript level
condition <- paste0(samples_new$condition)
counts.raw = trans_counts[rowSums(trans_counts>0)>0,]
counts.filtered = trans_counts[target_high$trans_high,]
mv.trans <- check.mean.variance(counts.raw = counts.raw,
                                counts.filtered = counts.filtered,
                                condition = condition)
### make plot
fit.raw <- mv.trans$fit.raw
fit.filtered <- mv.trans$fit.filtered
par(mfrow=c(1,2))
plot.mean.variance(x = fit.raw$sx,y = fit.raw$sy,
                   l = fit.raw$l,lwd=2,fit.line.col ='gold',col='black')
title('\n\nRaw counts (transcript level)')
plot.mean.variance(x = fit.filtered$sx,y = fit.filtered$sy,
                   l = fit.filtered$l,lwd=2,col='black')
title('\n\nFiltered counts (transcript level)')
lines(fit.raw$l, col = "gold",lty=4,lwd=2)
legend('topright',col = c('red','gold'),lty=c(1,4),lwd=3,
       legend = c('low-exp removed','low-exp kept'))

################################################################################
## gene level
condition <- paste0(samples_new$condition)
counts.raw = genes_counts[rowSums(genes_counts>0)>0,]
counts.filtered = genes_counts[target_high$genes_high,]
mv.genes <- check.mean.variance(counts.raw = counts.raw,
                                counts.filtered = counts.filtered,
                                condition = condition)
### make plot
fit.raw <- mv.genes$fit.raw
fit.filtered <- mv.genes$fit.filtered
par(mfrow=c(1,2))
plot.mean.variance(x = fit.raw$sx,y = fit.raw$sy,
                   l = fit.raw$l,lwd=2,fit.line.col ='gold',col='black')
title('\n\nRaw counts (gene level)')
plot.mean.variance(x = fit.filtered$sx,y = fit.filtered$sy,
                   l = fit.filtered$l,lwd=2,col='black')
title('\n\nFiltered counts (gene level)')
lines(fit.raw$l, col = "gold",lty=4,lwd=2)
legend('topright',col = c('red','gold'),lty=c(1,4),lwd=3,
       legend = c('low-exp removed','low-exp kept'))

```

### Step 3: Principal component analysis (PCA)
The PCA plot is a rough visualization of the expression variation across conditions/samples and to determine whether the data has batch effects between different biological replicates.
```{r}
################################################################################
##----->> trans level
data2pca <- trans_counts[target_high$trans_high,]
dge <- DGEList(counts=data2pca) 
dge <- calcNormFactors(dge)
data2pca <- t(counts2CPM(obj = dge,Log = T))
dim1 <- 'PC1'
dim2 <- 'PC2'
ellipse.type <- 'polygon' #ellipse.type=c('none','ellipse','polygon')

##--All Bio-reps plots
rownames(data2pca) <- gsub('_','.',rownames(data2pca))
groups <- samples_new$brep ## colour on biological replicates
# groups <- samples_new$condition ## colour on condtions
g <- plot.PCA.ind(data2pca = data2pca,dim1 = dim1,dim2 = dim2,
                  groups = groups,plot.title = 'Transcript PCA: bio-reps',
                  ellipse.type = ellipse.type,
                  add.label = T,adj.label = F)

g

##--average expression plot
rownames(data2pca) <- gsub('_','.',rownames(data2pca))
groups <- samples_new$brep
data2pca.ave <- rowmean(data2pca,samples_new$condition,reorder = F)
groups <- unique(samples_new$condition)
g <- plot.PCA.ind(data2pca = data2pca.ave,dim1 = 'PC1',dim2 = 'PC2',
                  groups = groups,plot.title = 'Transcript PCA: average expression',
                  ellipse.type = 'none',add.label = T,adj.label = F)

g

################################################################################
##----->> genes level
data2pca <- genes_counts[target_high$genes_high,]
dge <- DGEList(counts=data2pca) 
dge <- calcNormFactors(dge)
data2pca <- t(counts2CPM(obj = dge,Log = T))
dim1 <- 'PC1'
dim2 <- 'PC2'
ellipse.type <- 'polygon' #ellipse.type=c('none','ellipse','polygon')

##--All Bio-reps plots
rownames(data2pca) <- gsub('_','.',rownames(data2pca))
groups <- samples_new$brep ## colour on biological replicates
# groups <- samples_new$condition ## colour on condtions
g <- plot.PCA.ind(data2pca = data2pca,dim1 = dim1,dim2 = dim2,
                  groups = groups,plot.title = 'genescript PCA: bio-reps',
                  ellipse.type = ellipse.type,
                  add.label = T,adj.label = F)

g

##--average expression plot
rownames(data2pca) <- gsub('_','.',rownames(data2pca))
groups <- samples_new$brep
data2pca.ave <- rowmean(data2pca,samples_new$condition,reorder = F)
groups <- unique(samples_new$condition)
g <- plot.PCA.ind(data2pca = data2pca.ave,dim1 = 'PC1',dim2 = 'PC2',
                  groups = groups,plot.title = 'genescript PCA: average expression',
                  ellipse.type = 'none',add.label = T,adj.label = F)

g
```

### Step 3-continued: Batch effect estimation
Only if see clear batch effects on the PCA plots, otherwise skip this step. The <a href="https://bioconductor.org/packages/release/bioc/html/RUVSeq.html" target="_blank">RUVSeq</a> is used to estimate the batch effects. It will generate modified read counts, where the batch effects have been removed, for batch-free PCA plots and batch effect term which can be passed to the linear regression model for 3D analysis.

```{r}
ruvseq.method <- 'RUVr' # ruvseq.method is one of 'RUVr', 'RUVs' and 'RUVg'
design <- condition2design(condition = samples_new$condition,
                           batch.effect = NULL)


################################################################################
##----->> trans level
trans_batch <- remove.batch(read.counts = trans_counts[target_high$trans_high,],
                            condition = samples_new$condition,
                            design = design,
                            contrast=NULL,
                            group = samples_new$condition,
                            method = ruvseq.method)
save(trans_batch,file=paste0(data.folder,'/trans_batch.RData')) 

################################################################################
##----->> genes level
genes_batch <- remove.batch(read.counts = genes_counts[target_high$genes_high,],
                            condition = samples_new$condition,
                            design = design,
                            contrast=NULL,
                            group = samples_new$condition,
                            method = ruvseq.method)
save(genes_batch,file=paste0(data.folder,'/genes_batch.RData')) 


################################################################################
## DO the PCA again
################################################################################

##----->> trans level
data2pca <- trans_batch$normalizedCounts[target_high$trans_high,]
dge <- DGEList(counts=data2pca) 
dge <- calcNormFactors(dge)
data2pca <- t(counts2CPM(obj = dge,Log = T))
dim1 <- 'PC1'
dim2 <- 'PC2'
ellipse.type <- 'polygon' #ellipse.type=c('none','ellipse','polygon')

##--All Bio-reps plots
rownames(data2pca) <- gsub('_','.',rownames(data2pca))
groups <- samples_new$brep ## colour on biological replicates
# groups <- samples_new$condition ## colour on condtions
g <- plot.PCA.ind(data2pca = data2pca,dim1 = dim1,dim2 = dim2,
                  groups = groups,plot.title = 'Transcript PCA: bio-reps',
                  ellipse.type = ellipse.type,
                  add.label = T,adj.label = F)

g

##--average expression plot
rownames(data2pca) <- gsub('_','.',rownames(data2pca))
groups <- samples_new$brep
data2pca.ave <- rowmean(data2pca,samples_new$condition,reorder = F)
groups <- unique(samples_new$condition)
g <- plot.PCA.ind(data2pca = data2pca.ave,dim1 = 'PC1',dim2 = 'PC2',
                  groups = groups,plot.title = 'Transcript PCA: average expression',
                  ellipse.type = 'none',add.label = T,adj.label = F)

g

################################################################################
##----->> genes level
data2pca <- genes_batch$normalizedCounts[target_high$genes_high,]
dge <- DGEList(counts=data2pca) 
dge <- calcNormFactors(dge)
data2pca <- t(counts2CPM(obj = dge,Log = T))
dim1 <- 'PC1'
dim2 <- 'PC2'
ellipse.type <- 'polygon' #ellipse.type=c('none','ellipse','polygon')

##--All Bio-reps plots
rownames(data2pca) <- gsub('_','.',rownames(data2pca))
groups <- samples_new$brep ## colour on biological replicates
# groups <- samples_new$condition ## colour on condtions
g <- plot.PCA.ind(data2pca = data2pca,dim1 = dim1,dim2 = dim2,
                  groups = groups,plot.title = 'genescript PCA: bio-reps',
                  ellipse.type = ellipse.type,
                  add.label = T,adj.label = F)

g

##--average expression plot
rownames(data2pca) <- gsub('_','.',rownames(data2pca))
groups <- samples_new$brep
data2pca.ave <- rowmean(data2pca,samples_new$condition,reorder = F)
groups <- unique(samples_new$condition)
g <- plot.PCA.ind(data2pca = data2pca.ave,dim1 = 'PC1',dim2 = 'PC2',
                  groups = groups,plot.title = 'genescript PCA: average expression',
                  ellipse.type = 'none',add.label = T,adj.label = F)

g

```

### Step 4: Data normalisation
The same transcript/gene in deeplier sequenced samples have more mapped reads. To make genes/transcripts comparable across samples, the "TMM", "RLE" or "upperquantile" method can be used to normalize the expression.

```{r}
################################################################################
##----->> trans level
norm.method <- 'TMM' ## norm.method is one of 'TMM','RLE' and 'upperquartile'
dge <- DGEList(counts=trans_counts[target_high$trans_high,],
               group = samples_new$condition,
               genes = mapping[target_high$trans_high,])
trans_dge <- suppressWarnings(calcNormFactors(dge,method = norm.method))
save(trans_dge,file=paste0(data.folder,'/trans_dge.RData'))

################################################################################
##----->> genes level
dge <- DGEList(counts=genes_counts[target_high$genes_high,],
               group = samples_new$condition)
genes_dge <- suppressWarnings(calcNormFactors(dge,method = norm.method))
save(genes_dge,file=paste0(data.folder,'/genes_dge.RData'))

################################################################################
##----->> distribution plot
sample.name <- paste0(samples_new$condition,'.',samples_new$brep)
condition <- samples_new$condition

###--- trans level
data.before <- trans_counts[target_high$trans_high,]
data.after <- counts2CPM(obj = trans_dge,Log = T)
g <- boxplot.normalised(data.before = data.before,
                        data.after = data.after,
                        condition = condition,
                        sample.name = sample.name)
do.call(grid.arrange,g)

###--- genes level
data.before <- genes_counts[target_high$genes_high,]
data.after <- counts2CPM(obj = genes_dge,Log = T)
g <- boxplot.normalised(data.before = data.before,
                        data.after = data.after,
                        condition = condition,
                        sample.name = sample.name)
do.call(grid.arrange,g)
```


## DE, DAS and DTU analysis
Three pipelines, <a href="https://bioconductor.org/packages/release/bioc/html/limma.html" target="_blank">limma</a>, <a href="https://bioconductor.org/packages/release/bioc/html/edgeR.html" target="_blank">glmQL (edgeR)</a> and <a href="https://bioconductor.org/packages/release/bioc/html/edgeR.html" target="_blank">glm (edgeR)</a> are provided for DE gene and transcripts analysis. "limma" and "edgeR" have comparable performance and they have strigent controls of false positives than the "glm".

### Step 1: Load contrast group
The contrast groups passed to the analysis are in a vector object, e.g. \code{contrast <- c('B-A','C-A')} means compare conditions "B" and "C" to condition "A".

```{r}
contrast <- read.csv(paste0(input.folder,'/contrast.csv'))
contrast <- paste0(contrast[,1],'-',contrast[,2])
###---generate proper contrast format
contrast <- gsub('_','.',contrast)
```

### Step 2: DE genes
```{r}
batch.effect <- genes_batch$W
# batch.effect <- NULL ## if has no batch effects
design <- condition2design(condition = samples_new$condition,
                           batch.effect = batch.effect)

################################################################################
p.adjust.method <- 'BH'
pval.cutoff <- 0.01
lfc.cutoff <- 1
##----->> limma pipeline
genes_3D_stat <- limma.pipeline(dge = genes_dge,
                                 design = design,
                                 deltaPS = NULL,
                                 contrast = contrast,
                                 diffAS = F,
                                 adjust.method = p.adjust.method)

##----->> edgeR glmQL pipeline
# genes_3D_stat <- edgeR.pipeline(dge = genes_dge,
#                                  design = design,
#                                  deltaPS = NULL,
#                                  contrast = contrast,
#                                  diffAS = F,
#                                  method = 'glmQL',
#                                  adjust.method = p.adjust.method)

##----->> edgeR glm pipeline
# genes_3D_stat <- edgeR.pipeline(dge = genes_dge,
#                                  design = design,
#                                  deltaPS = NULL,
#                                  contrast = contrast,
#                                  diffAS = F,
#                                  method = 'glm',
#                                  adjust.method = p.adjust.method)

## save results
save(genes_3D_stat,file=paste0(data.folder,'/genes_3D_stat.RData'))

```

### Step 3: DAS genes, DE and DTU transcripts

```{r}
################################################################################
##----->> generate deltaPS
transAbundance <- txi_trans$abundance[target_high$trans_high,]
deltaPS <- transAbundance2PS(transAbundance = transAbundance,
                             PS = NULL,
                             contrast = contrast,
                             condition = samples$condition,
                             mapping = mapping[target_high$trans_high,])
PS <- deltaPS$PS
deltaPS <- deltaPS$deltaPS
## save 
save(deltaPS,file=paste0(data.folder,'/deltaPS.RData')) 
save(PS,file=paste0(data.folder,'/PS.RData'))

################################################################################
##----->> DAS genes,DE and DTU transcripts
batch.effect <- genes_batch$W
# batch.effect <- NULL ## if has no batch effects
design <- condition2design(condition = samples_new$condition,
                           batch.effect = batch.effect)

################################################################################
p.adjust.method <- 'BH'
pval.cutoff <- 0.01
lfc.cutoff <- 1
deltaPS.cutoff <- 0.1
##----->> limma pipeline
trans_3D_stat <- limma.pipeline(dge = trans_dge,
                                 design = design,
                                 deltaPS = deltaPS,
                                 contrast = contrast,
                                 diffAS = T,
                                 adjust.method = p.adjust.method)

##----->> edgeR glmQL pipeline
# trans_3D_stat <- edgeR.pipeline(dge = trans_dge,
#                                  design = design,
#                                  deltaPS = deltaPS,
#                                  contrast = contrast,
#                                  diffAS = T,
#                                  method = 'glmQL',
#                                  adjust.method = p.adjust.method)

##----->> edgeR glm pipeline
# trans_3D_stat <- edgeR.pipeline(dge = trans_dge,
#                                  design = design,
#                                  deltaPS = deltaPS,
#                                  contrast = contrast,
#                                  diffAS = T,
#                                  method = 'glm',
#                                  adjust.method = p.adjust.method)

## save results
save(trans_3D_stat,file=paste0(data.folder,'/trans_3D_stat.RData'))

```

## Result summary
### 3D target and tables of statistics
```{r}
################################################################################
##----->> Summary DE genes
DE_genes <- summary.DE.target(stat = genes_3D_stat$DE.stat,
                              cutoff = c(adj.pval=pval.cutoff,
                                         log2FC=lfc.cutoff))

################################################################################
## summary DAS genes, DE and DTU trans
##----->> DE trans
DE_trans <- summary.DE.target(stat = trans_3D_stat$DE.stat,
                              cutoff = c(adj.pval=pval.cutoff,
                                         log2FC=lfc.cutoff))

##----->> DAS genes
DAS.p.method <- 'F-test' ## DAS.p.method is one of 'F-test' and 
if(DAS.p.method=='F-test') {
  DAS.stat <- trans_3D_stat$DAS.F.stat
} else {
  DAS.stat <- trans_3D_stat$DAS.Simes.stat
}

lfc <- genes_3D_stat$DE.lfc
lfc <- reshape2::melt(as.matrix(lfc))
colnames(lfc) <- c('target','contrast','log2FC')
DAS_genes <- summary.DAS.target(stat = DAS.stat,
                                lfc = lfc,
                                cutoff=c(pval.cutoff,deltaPS.cutoff))

##----->> DTU trans
lfc <- trans_3D_stat$DE.lfc
lfc <- reshape2::melt(as.matrix(lfc))
colnames(lfc) <- c('target','contrast','log2FC')
DTU_trans <- summary.DAS.target(stat = trans_3D_stat$DTU.stat,
                                lfc = lfc,cutoff = c(adj.pval=pval.cutoff,
                                                     deltaPS=deltaPS.cutoff))

################################################################################
## save results
save(DE_genes,file=paste0(data.folder,'/DE_genes.RData'))
save(DE_trans,file=paste0(data.folder,'/DE_trans.RData')) 
save(DAS_genes,file=paste0(data.folder,'/DAS_genes.RData')) 
save(DTU_trans,file=paste0(data.folder,'/DTU_trans.RData')) 

## csv
write.csv(DE_genes,file=paste0(result.folder,'/DE genes.csv'),row.names = F)
write.csv(DAS_genes,file=paste0(result.folder,'/DAS genes.csv'),row.names = F)
write.csv(DE_trans,file=paste0(result.folder,'/DE transcripts.csv'),row.names = F)
write.csv(DTU_trans,file=paste0(result.folder,'/DTU transcripts.csv'),row.names = F)

```

### 3D numbers
```{r}
################################################################################
##----->> target numbers
summary.3D.number(DE_genes = DE_genes,
                   DAS_genes = DAS_genes,
                   DE_trans = DE_trans,
                   DTU_trans=DTU_trans,
                   contrast = contrast)

################################################################################
##----->> DE vs DAS
DEvsDAS(DE_genes = DE_genes,
        DAS_genes = DAS_genes,
        contrast = contrast)


################################################################################
##----->> DE vs DTU
DEvsDTU(DE_trans = DE_trans,
        DTU_trans = DTU_trans,
        contrast = contrast)

```

### Make plot

#### Up- and down-regulation
```{r}
################################################################################
##----->> DE genes
idx <- factor(DE_genes$contrast,levels = contrast)
targets <-  split(DE_genes,idx)
data2plot <- lapply(contrast,function(i){
  if(nrow(targets[[i]])==0){
    x <- data.frame(contrast=i,regulation=c('down_regulate','up_regulate'),number=0)
  } else {
    x <- data.frame(contrast=i,table(targets[[i]]$up.down))
    colnames(x) <- c('contrast','regulation','number')
  }
  x
})
data2plot <- do.call(rbind,data2plot)
plot.updown(data2plot,plot.title = 'DE genes',contrast = contrast)

################################################################################
##----->> DAS genes
idx <- factor(DAS_genes$contrast,levels = contrast)
targets <-  split(DAS_genes,idx)
data2plot <- lapply(contrast,function(i){
  if(nrow(targets[[i]])==0){
    x <- data.frame(contrast=i,regulation=c('down_regulate','up_regulate'),number=0)
  } else {
    x <- data.frame(contrast=i,table(targets[[i]]$up.down))
    colnames(x) <- c('contrast','regulation','number')
  }
  x
})
data2plot <- do.call(rbind,data2plot)
plot.updown(data2plot,plot.title = 'DAS genes',contrast = contrast)

################################################################################
##----->> DE trans
idx <- factor(DE_trans$contrast,levels = contrast)
targets <-  split(DE_trans,idx)
data2plot <- lapply(contrast,function(i){
  if(nrow(targets[[i]])==0){
    x <- data.frame(contrast=i,regulation=c('down_regulate','up_regulate'),number=0)
  } else {
    x <- data.frame(contrast=i,table(targets[[i]]$up.down))
    colnames(x) <- c('contrast','regulation','number')
  }
  x
})
data2plot <- do.call(rbind,data2plot)
plot.updown(data2plot,plot.title = 'DE trans',contrast = contrast)

################################################################################
##----->> DTU trans
idx <- factor(DTU_trans$contrast,levels = contrast)
targets <-  split(DTU_trans,idx)
data2plot <- lapply(contrast,function(i){
  if(nrow(targets[[i]])==0){
    x <- data.frame(contrast=i,regulation=c('down_regulate','up_regulate'),number=0)
  } else {
    x <- data.frame(contrast=i,table(targets[[i]]$up.down))
    colnames(x) <- c('contrast','regulation','number')
  }
  x
})
data2plot <- do.call(rbind,data2plot)
plot.updown(data2plot,plot.title = 'DTU trans',contrast = contrast)


```

#### Flow chart
```{r}
################################################################################
##----->> DE vs DAS genes
DE.genes <- unique(DE_genes$target) 
DAS.genes <- unique(DAS_genes$target) 
plot.flow.chart(expressed = target_high$genes_high, 
                x = DE.genes, 
                y = DAS.genes, 
                type = 'genes', 
                pval.cutoff = pval.cutoff,
                lfc.cutoff = lfc.cutoff,
                deltaPS.cutoff = deltaPS.cutoff)

################################################################################
##----->> DE vs DTU transcripts
DE.trans<- unique(DE_trans$target)
DTU.trans <- unique(DTU_trans$target)

plot.flow.chart(expressed = target_high$trans_high, 
                x = DE.trans, 
                y = DTU.trans, 
                type = 'transcripts', 
                pval.cutoff = pval.cutoff,
                lfc.cutoff = lfc.cutoff,
                deltaPS.cutoff = deltaPS.cutoff)

```

#### Plot compare contrast groups
```{r}
################################################################################
##----->> DE genes
targets <- lapply(contrast,function(i){
  subset(DE_genes,contrast==i)$target
})
names(targets) <- contrast
g <- plot.euler.diagram(x = targets)
grid.arrange(g,top=textGrob('DE genes', gp=gpar(cex=1.2)))
   
################################################################################
##----->> DAS genes
targets <- lapply(contrast,function(i){
  subset(DAS_genes,contrast==i)$target
})
names(targets) <- contrast
g <- plot.euler.diagram(x = targets)
grid.arrange(g,top=textGrob('DAS genes', gp=gpar(cex=1.2)))

################################################################################
##----->> DE transcripts
targets <- lapply(contrast,function(i){
  subset(DE_trans,contrast==i)$target
})
names(targets) <- contrast
g <- plot.euler.diagram(x = targets)
grid.arrange(g,top=textGrob('DE transcripts', gp=gpar(cex=1.2)))
    
################################################################################
##----->> DTU transcripts
targets <- lapply(contrast,function(i){
  subset(DTU_trans,contrast==i)$target
})
names(targets) <- contrast
g <- plot.euler.diagram(x = targets)
grid.arrange(g,top=textGrob('DTU transcripts', gp=gpar(cex=1.2)))
   
```

#### Plot transcriptional vs AS
```{r}
contrast.idx <- contrast[1]
################################################################################
##----->> DE vs DAS genes
x <- unlist(DEvsDAS[DEvsDAS$Contrast==contrast.idx,-1])
if(length(x)==0){
  message('No DE and/or DAS genes')
} else {
  names(x) <- c('DE','DE&DAS','DAS')
  g <- plot.euler.diagram(x = x,fill = gg.color.hue(2))
  g
  grid.arrange(g,top=textGrob('DE vs DAS genes', gp=gpar(cex=1.2)))
}

################################################################################
##----->> DE vs DTU transcripts
x <- unlist(DEvsDTU[DEvsDTU$Contrast==contrast.idx,-1])
if(length(x)==0){
  message('No DE and/or DTU transcripts')
} else {
  names(x) <- c('DE','DE&DTU','DTU')
  g <- plot.euler.diagram(x = x,fill = gg.color.hue(2))
  g
  grid.arrange(g,top=textGrob('DE vs DTU transcripts', gp=gpar(cex=1.2)))
}
```

## Advanced plot
### Heatmap 
```{r}
################################################################################
##----->> DE genes
dist.method <- 'euclidean'
cluster.method <- 'ward.D'
cluster.number <- 10
targets <- unique(DE_genes$target)
data2heatmap <- txi_genes$abundance[targets,]
column_title <- paste0(length(targets),' DE genes')
data2plot <- rowmean(x = t(data2heatmap),
                     group = samples$condition,
                     reorder = F)
data2plot <- t(scale(data2plot))

hc.dist <- dist(data2plot,method = dist.method)
hc <- fastcluster::hclust(hc.dist,method = cluster.method)
clusters <- cutree(hc, k = cluster.number)
clusters <- reorder.clusters(clusters = clusters,dat = data2plot)

g <- Heatmap(as.matrix(data2plot), name = 'Z-scores', 
             cluster_rows = TRUE,
             clustering_method_rows=cluster.method,
             row_dend_reorder = T,
             show_row_names = FALSE, 
             show_column_names = ifelse(ncol(data2plot)>10,F,T),
             cluster_columns = FALSE,
             split=clusters,
             column_title= column_title)

draw(g,column_title='Conditions',column_title_side = "bottom")

################################################################################
##----->> DAS genes
dist.method <- 'euclidean'
cluster.method <- 'ward.D'
cluster.number <- 10
targets <- intersect(unique(DE_genes$target),unique(DAS_genes$target))
data2heatmap <- txi_genes$abundance[targets,]
column_title <- paste0(length(targets),' DE&DAS genes')
data2plot <- rowmean(x = t(data2heatmap),
                     group = samples$condition,
                     reorder = F)
data2plot <- t(scale(data2plot))

hc.dist <- dist(data2plot,method = dist.method)
hc <- fastcluster::hclust(hc.dist,method = cluster.method)
clusters <- cutree(hc, k = cluster.number)
clusters <- reorder.clusters(clusters = clusters,dat = data2plot)

g <- Heatmap(as.matrix(data2plot), name = 'Z-scores', 
             cluster_rows = TRUE,
             clustering_method_rows=cluster.method,
             row_dend_reorder = T,
             show_row_names = FALSE, 
             show_column_names = ifelse(ncol(data2plot)>10,F,T),
             cluster_columns = FALSE,
             split=clusters,
             column_title= column_title)

draw(g,column_title='Conditions',column_title_side = "bottom")

################################################################################
##----->> DE trans
dist.method <- 'euclidean'
cluster.method <- 'ward.D'
cluster.number <- 10
targets <- unique(DE_trans$target)
data2heatmap <- txi_trans$abundance[targets,]
column_title <- paste0(length(targets),' DE trans')
data2plot <- rowmean(x = t(data2heatmap),
                     group = samples$condition,
                     reorder = F)
data2plot <- t(scale(data2plot))

hc.dist <- dist(data2plot,method = dist.method)
hc <- fastcluster::hclust(hc.dist,method = cluster.method)
clusters <- cutree(hc, k = cluster.number)
clusters <- reorder.clusters(clusters = clusters,dat = data2plot)

g <- Heatmap(as.matrix(data2plot), name = 'Z-scores', 
             cluster_rows = TRUE,
             clustering_method_rows=cluster.method,
             row_dend_reorder = T,
             show_row_names = FALSE, 
             show_column_names = ifelse(ncol(data2plot)>10,F,T),
             cluster_columns = FALSE,
             split=clusters,
             column_title= column_title)

draw(g,column_title='Conditions',column_title_side = "bottom")

################################################################################
##----->> DTU trans
dist.method <- 'euclidean'
cluster.method <- 'ward.D'
cluster.number <- 10
targets <- intersect(unique(DE_trans$target),unique(DTU_trans$target))
data2heatmap <- txi_trans$abundance[targets,]
column_title <- paste0(length(targets),' DE&DTU trans')
data2plot <- rowmean(x = t(data2heatmap),
                     group = samples$condition,
                     reorder = F)
data2plot <- t(scale(data2plot))

hc.dist <- dist(data2plot,method = dist.method)
hc <- fastcluster::hclust(hc.dist,method = cluster.method)
clusters <- cutree(hc, k = cluster.number)
clusters <- reorder.clusters(clusters = clusters,dat = data2plot)

g <- Heatmap(as.matrix(data2plot), name = 'Z-scores', 
             cluster_rows = TRUE,
             clustering_method_rows=cluster.method,
             row_dend_reorder = T,
             show_row_names = FALSE, 
             show_column_names = ifelse(ncol(data2plot)>10,F,T),
             cluster_columns = FALSE,
             split=clusters,
             column_title= column_title)

draw(g,column_title='Conditions',column_title_side = "bottom")

```

### Profile plot
```{r}
################################################################################
##----->> generate annotation of DE and/or DAS, DE and/or DTU
DE.genes <- unique(DE_genes$target)
DAS.genes <- unique(DAS_genes$target)
DE.trans <- unique(DE_trans$target)
DTU.trans <- unique(DTU_trans$target)

genes.ann <- set2(DE.genes,DAS.genes)
names(genes.ann) <- c('DEonly','DE&DAS','DASonly')
genes.ann <- plyr::ldply(genes.ann,cbind)[,c(2,1)]
colnames(genes.ann) <- c('target','annotation')


trans.ann <- set2(DE.trans,DTU.trans)
names(trans.ann) <- c('DEonly','DE&DTU','DTUonly')
trans.ann <- plyr::ldply(trans.ann,cbind)[,c(2,1)]
colnames(trans.ann) <- c('target','annotation')

################################################################################
##give a gene name
gene <- 'AT5G24470'
##----->> profile plot of TPM or read counts
g.pr <- plot.abundance(data.exp = txi_trans$abundance[target_high$trans_high,],
                       gene = gene,
                       mapping = mapping[target_high$trans_high,],
                       genes.ann = genes.ann,
                       trans.ann = trans.ann,
                       trans.expressed = NULL,
                       reps = samples$condition,
                       y.lab = 'TPM')
g.pr

################################################################################
##----->> PS plot
g.ps <- plot.PS(data.exp = txi_trans$abundance[target_high$trans_high,],
                gene = gene,
                mapping = mapping[target_high$trans_high,],
                genes.ann = genes.ann,
                trans.ann = trans.ann,
                trans.expressed = NULL,
                reps = samples$condition,
                y.lab = 'PS')
g.ps

```

### GO annotation plot
```{r}
################################################################################
##----->> DE genes
go.table <- suppressMessages(readr::read_csv(paste0(input.folder,'/DE genes GO terms.csv')))
go.table <- go.table[,c(1,2,5)]
plotGO(go.table = go.table,col.idx = '-log10(FDR)',
       plot.title = 'GO annotation: DE genes')

################################################################################
##----->> DAS genes
go.table <- suppressMessages(readr::read_csv(paste0(input.folder,'/DAS genes GO terms.csv')))
go.table <- go.table[,c(1,2,5)]
plotGO(go.table = go.table,col.idx = '-log10(FDR)',
       plot.title = 'GO annotation: DAS genes')

```
